File: src/mvmctl/core/vm/_controller.py
"""
VM lifecycle operations.

This module contains the VMController class for managing VM lifecycle operations
like start, stop, pause, resume, ssh, logs, etc.
"""

from __future__ import annotations

import logging
from pathlib import Path

from mvmctl.core.vm._firecracker import FirecrackerClient
from mvmctl.core.vm._repository import VMRepository
from mvmctl.exceptions import MVMError, VMStateError
from mvmctl.models import VMInstanceItem, VMStatus
from mvmctl.models.volume import VolumeItem
from mvmctl.utils._system import ProcessSignalHandler

logger = logging.getLogger(__name__)


class VMController:
    """
    Stateful VM lifecycle manager.

    Resolves VM entity in __init__ and operates on cached VM instance.
    """

    def __init__(
        self,
        entity: str | VMInstanceItem,
        repo: VMRepository,
    ) -> None:
        from mvmctl.core.vm._resolver import VMResolver

        self._repo = repo

        if isinstance(entity, VMInstanceItem):
            self._vm = entity
        else:
            self._resolver = VMResolver(self._repo)
            self._vm = self._resolver.resolve(entity)

    def stop(self, force: bool = False) -> None:
        """
        Stop the VM (idempotent — never raises).

        If the VM is already stopped or the underlying process is gone,
        returns immediately.  If the process exists but cannot be stopped,
        the status is set to ERROR and the method returns cleanly so
        that removal cleanup can still proceed.
        """

        handler = ProcessSignalHandler(
            self._vm.pid,
            expected_start_time=self._vm.process_start_time,
        )

        # ── Non-running VMs: idempotent + orphan cleanup ──
        # The DB status might be STOPPED/PAUSED/ERROR but the actual
        # firecracker process could still be running (e.g., after a
        # failed cleanup or orphaned process from a previous run).
        if self._vm.status not in (
            VMStatus.RUNNING.value,
            VMStatus.STARTING.value,
        ):
            if self._vm.pid and handler.is_alive():
                handler.kill()
                self._repo.update_status(self._vm.id, VMStatus.STOPPED.value)
                self._vm.status = VMStatus.STOPPED.value
            return

        # ── RUNNING/STARTING but process is already gone ──
        if not self._vm.pid or not handler.is_alive():
            self._repo.update_status(self._vm.id, VMStatus.STOPPED.value)
            return

        # ── Normal stop: process is alive, VM is running ──
        self._repo.update_status(self._vm.id, VMStatus.STOPPING.value)
        try:
            if not force and self._vm.api_socket_path:
                # Try graceful shutdown via Firecracker API first
                try:
                    client = FirecrackerClient(
                        self._vm.vm_dir / self._vm.api_socket_path
                    )
                    client.send_ctrl_alt_del()
                    client.close()
                    # Wait for guest OS shutdown via pre_signal_hook
                    exit_code = handler.graceful_shutdown(
                        pre_signal_hook=lambda: False,
                    )
                except Exception:
                    # Fall through to signal-based shutdown
                    exit_code = None

                if exit_code is None:
                    exit_code = handler.graceful_shutdown()
            else:
                if force:
                    handler.kill()
                exit_code = handler.graceful_shutdown()

            # Capture exit code if not already captured
            if exit_code is None:
                exit_code = handler.wait_and_capture_exit()

            # Persist exit code to database
            if exit_code is not None:
                self._repo.update_exit_code(self._vm.id, exit_code)

            self._repo.update_status(self._vm.id, VMStatus.STOPPED.value)
        except Exception as exc:
            self._repo.update_status(self._vm.id, VMStatus.ERROR.value)
            logger.warning("Failed to stop VM '%s': %s", self._vm.name, exc)

    def pause(self) -> None:
        """
        Pause the VM (idempotent — no-op if already paused).

        Raises:
            MVMError: If VM cannot be paused from its current state

        """
        name = self._vm.name

        # No-op — already paused
        if self._vm.status == VMStatus.PAUSED.value:
            return

        # Cannot pause from these states
        if self._vm.status == VMStatus.STARTING.value:
            raise VMStateError(
                f"VM '{name}' is still starting — cannot pause (current state: {self._vm.status})"
            )
        if self._vm.status == VMStatus.STOPPED.value:
            raise VMStateError(
                f"VM '{name}' is stopped — cannot pause (current state: {self._vm.status})"
            )
        if self._vm.status == VMStatus.STOPPING.value:
            raise VMStateError(
                f"VM '{name}' is shutting down — cannot pause (current state: {self._vm.status})"
            )
        if self._vm.status in (VMStatus.ERROR.value, VMStatus.CRASHED.value):
            raise VMStateError(
                f"VM '{name}' is in {self._vm.status} state — cannot pause (current state: {self._vm.status})"
            )

        # Valid transition — must be RUNNING
        if not self._vm.api_socket_path:
            raise VMStateError(f"VM '{name}' has no API socket enabled")
        client = FirecrackerClient(self._vm.vm_dir / self._vm.api_socket_path)
        try:
            client.pause_vm()
            self._repo.update_status(self._vm.id, VMStatus.PAUSED.value)
        finally:
            client.close()

    def resume(self) -> None:
        """
        Resume the VM (idempotent — no-op if already running/starting).

        Raises:
            MVMError: If VM cannot be resumed from its current state

        """
        name = self._vm.name

        # No-op — already in or moving toward target state (RUNNING)
        if self._vm.status in (VMStatus.RUNNING.value, VMStatus.STARTING.value):
            return

        # Error/crashed state
        if self._vm.status in (VMStatus.ERROR.value, VMStatus.CRASHED.value):
            raise VMStateError(
                f"VM '{name}' is in {self._vm.status} state — remove and recreate (current state: {self._vm.status})"
            )

        # Wrong direction — stopped
        if self._vm.status == VMStatus.STOPPED.value:
            raise VMStateError(
                f"VM '{name}' is stopped — use start() instead (current state: {self._vm.status})"
            )

        # Wrong direction — shutting down
        if self._vm.status == VMStatus.STOPPING.value:
            raise VMStateError(
                f"VM '{name}' is shutting down — use start() after it stops (current state: {self._vm.status})"
            )

        # Valid transition — must be PAUSED
        if not self._vm.api_socket_path:
            raise VMStateError(f"VM '{name}' has no API socket enabled")
        client = FirecrackerClient(self._vm.vm_dir / self._vm.api_socket_path)
        try:
            client.resume_vm()
            self._repo.update_status(self._vm.id, VMStatus.RUNNING.value)
        finally:
            client.close()

    def start(self) -> None:
        """
        Start an already configured VM via Firecracker API (idempotent — no-op if
        already running, starting, or stopping).

        Raises:
            MVMError: If VM cannot be started from its current state

        """
        name = self._vm.name

        # No-op — already in or moving toward target state (RUNNING),
        # or will be stopped soon (retry start after)
        if self._vm.status in (
            VMStatus.RUNNING.value,
            VMStatus.STARTING.value,
            VMStatus.STOPPING.value,
        ):
            return

        # Error/crashed state
        if self._vm.status in (VMStatus.ERROR.value, VMStatus.CRASHED.value):
            raise VMStateError(
                f"VM '{name}' is in {self._vm.status} state — remove and recreate (current state: {self._vm.status})"
            )

        # Wrong direction — paused
        if self._vm.status == VMStatus.PAUSED.value:
            raise VMStateError(
                f"VM '{name}' is paused — use resume() instead (current state: {self._vm.status})"
            )

        # Valid transition — must be STOPPED
        if not self._vm.api_socket_path:
            raise VMStateError(f"VM '{name}' has no API socket enabled")

        client = FirecrackerClient(self._vm.vm_dir / self._vm.api_socket_path)
        try:
            client.start_instance()
            self._repo.update_status(self._vm.id, VMStatus.RUNNING.value)
        finally:
            client.close()

    def reboot(self, force: bool = False) -> None:
        """
        Reboot the VM (stop then boot).

        Args:
            force: If True, force immediate shutdown

        Raises:
            MVMError: If reboot fails

        """
        self.stop(force=force)
        self.start()

    def snapshot(self, mem_out: Path, state_out: Path) -> None:
        """
        Snapshot VM memory and disk state.

        The VM is automatically paused before snapshotting and resumed
        afterwards so the caller gets back a running VM.

        Args:
            mem_out: Memory snapshot output path
            state_out: VM state output path

        Raises:
            MVMError: If socket not found, VM in invalid state, or
                      snapshot fails

        """
        name = self._vm.name

        # Validate state — snapshot requires RUNNING or PAUSED
        if self._vm.status == VMStatus.STARTING.value:
            raise VMStateError(
                f"VM '{name}' is still starting — cannot snapshot (current state: {self._vm.status})"
            )
        if self._vm.status == VMStatus.STOPPED.value:
            raise VMStateError(
                f"VM '{name}' is stopped — cannot snapshot (current state: {self._vm.status})"
            )
        if self._vm.status == VMStatus.STOPPING.value:
            raise VMStateError(
                f"VM '{name}' is shutting down — cannot snapshot (current state: {self._vm.status})"
            )
        if self._vm.status in (VMStatus.ERROR.value, VMStatus.CRASHED.value):
            raise VMStateError(
                f"VM '{name}' is in {self._vm.status} state — cannot snapshot (current state: {self._vm.status})"
            )

        if not self._vm.api_socket_path:
            raise VMStateError(
                f"Socket not found for VM '{name}'. Must be running with --enable-api-socket"
            )

        client = FirecrackerClient(self._vm.vm_dir / self._vm.api_socket_path)
        was_running = self._vm.status == VMStatus.RUNNING.value
        try:
            # Pause before snapshotting (Firecracker requires VM to be paused)
            if was_running:
                client.pause_vm()
                self._repo.update_status(self._vm.id, VMStatus.PAUSED.value)
                self._vm.status = VMStatus.PAUSED.value

            client.create_snapshot(mem_out, state_out)
        finally:
            # Resume if we paused it
            if was_running:
                try:
                    client.resume_vm()
                    self._repo.update_status(
                        self._vm.id, VMStatus.RUNNING.value
                    )
                    self._vm.status = VMStatus.RUNNING.value
                except MVMError:
                    logger.warning(
                        "Failed to resume VM '%s' after snapshot — "
                        "leaving in paused state",
                        name,
                    )
            client.close()

    def load_snapshot(
        self, mem_in: Path, state_in: Path, resume_after: bool = False
    ) -> None:
        """
        Load VM from snapshot.

        Args:
            mem_in: Memory snapshot input path
            state_in: VM state input path
            resume_after: Whether to resume VM after loading

        Raises:
            MVMError: If socket not found or load fails

        """
        if not self._vm.api_socket_path:
            raise VMStateError(
                f"Socket not found for VM '{self._vm.name}'. Must be running with --enable-api-socket"
            )
        client = FirecrackerClient(self._vm.vm_dir / self._vm.api_socket_path)
        try:
            client.load_snapshot(mem_in, state_in, resume_after)
            # Update status based on whether VM was resumed
            new_status = (
                VMStatus.RUNNING.value
                if resume_after
                else VMStatus.PAUSED.value
            )
            self._repo.update_status(self._vm.id, new_status)
            self._vm.status = new_status
        finally:
            client.close()

    def attach_volume(self, vol: VolumeItem) -> None:
        """Hotplug a drive into the running VM and persist to config.

        Args:
            vol: The volume to attach.

        """
        from mvmctl.core.vm._firecracker import FirecrackerConfigManager

        drive_config = {
            "drive_id": vol.id,
            "path_on_host": vol.path,
            "is_root_device": False,
            "is_read_only": False,
            "cache_type": "Unsafe",
            "io_engine": "Sync",
        }
        # Hotplug into the running Firecracker process
        sock_path = self._vm.vm_dir / self._vm.api_socket_path
        client = FirecrackerClient(sock_path)
        try:
            client.put_drive(drive_config)
        finally:
            client.close()
        # Persist to the Firecracker config JSON so it survives reboot
        config_mgr = FirecrackerConfigManager(
            self._vm.vm_dir / self._vm.config_path
        )
        config_mgr.add_drive(drive_config)
        logger.info(
            "Attached volume '%s' to VM '%s'",
            vol.id,
            self._vm.name,
        )

    def detach_volume(self, vol: VolumeItem) -> None:
        """Remove a drive from a running VM and update config.

        Args:
            vol: The volume to detach.

        """
        from mvmctl.core.vm._firecracker import FirecrackerConfigManager

        # Call Firecracker API to delete the drive
        sock_path = self._vm.vm_dir / self._vm.api_socket_path
        client = FirecrackerClient(sock_path)
        try:
            client.delete_drive(vol.id)
        finally:
            client.close()
        # Update Firecracker config JSON on disk
        FirecrackerConfigManager(
            self._vm.vm_dir / self._vm.config_path
        ).remove_drive(vol.id)
        logger.info(
            "Detached volume '%s' from VM '%s'",
            vol.id,
            self._vm.name,
        )


__all__ = ["VMController"]

File: internal/core/vm/controller.go
package vm

import (
	"context"
	"fmt"
	"log/slog"
	"time"

	"mvmctl/internal/lib/firecracker"
	"mvmctl/internal/lib/model"
	"mvmctl/internal/lib/system"
	"mvmctl/pkg/errs"
)

// --- Controller ---

// Controller manages per-VM lifecycle operations.
type Controller struct {
	vm   *model.VMItem
	repo Repository
}

// NewController creates a new VM controller.
func NewController(vm *model.VMItem, repo Repository) *Controller {
	return &Controller{vm: vm, repo: repo}
}

// --- Stop ---
//
// Stop is idempotent — never returns an error.
// If the VM is already stopped or the underlying process is gone,
// returns immediately. If the process exists but cannot be stopped,
// the status is set to ERROR and the method returns cleanly so
// that removal cleanup can still proceed.
//
// First two code paths (non-running, process-gone) let errors propagate.
func (c *Controller) Stop(ctx context.Context, force bool) error {
	pid := c.vm.PID

	// --- Non-running VMs: idempotent + orphan cleanup ---
	// The DB status might be STOPPED/PAUSED/ERROR but the actual
	// firecracker process could still be running (e.g., after a
	// failed cleanup or orphaned process from a previous run).
	//
	// Errors must propagate to caller — NOT absorbed.
	if c.vm.Status != model.VMStatusRunning && c.vm.Status != model.VMStatusStarting {
		if pid > 0 && system.IsProcessAlive(pid, c.vm.ProcessStartTime) {
			system.KillProcess(pid)
			if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusStopped); err != nil {
				return err
			}
			c.vm.Status = model.VMStatusStopped
		} else if pid > 0 {
			// Process is a zombie or dead — reap exit code if child
			system.CaptureExitCode(pid)
		}
		return nil
	}

	// --- RUNNING/STARTING but process is already gone ---
	// Does NOT update in-memory status here (only DB).
	if pid == 0 || !system.IsProcessAlive(pid, c.vm.ProcessStartTime) {
		if pid > 0 {
			// Process is a zombie or dead — reap exit code if child
			system.CaptureExitCode(pid)
		}
		if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusStopped); err != nil {
			return err
		}
		return nil
	}

	// --- Normal stop: process is alive, VM is running ---
	// Does NOT update in-memory status here (only DB).
	if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusStopping); err != nil {
		return err
	}

	stopErr := c.shutdownProcess(ctx, force, pid)
	if stopErr != nil {
		// Set status to ERROR and absorb the error.
		if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusError); err != nil {
			slog.Warn("Failed to update VM status to ERROR", "error", err)
		}
		slog.Warn("Failed to stop VM", "name", c.vm.Name, "error", stopErr)
		return nil
	}
	return nil
}

// shutdownProcess handles the actual Firecracker process shutdown.
// Any error returned causes Stop() to set DB status to ERROR.
func (c *Controller) shutdownProcess(ctx context.Context, force bool, pid int) error {
	var exitCode *int

	if !force && c.vm.APISocketPath != "" {
		apiSocket := c.vm.APISocketPath

		// Try graceful shutdown via Firecracker API first
		client := firecracker.NewClient(apiSocket)
		wasCtrlAltDel, ctrlErr := client.SendCtrlAltDel(ctx)
		client.Close()

		// If Ctrl+Alt+Del was sent, use hook-based wait: the guest OS
		// initiates shutdown, and the hook (returns false) tells the
		// signal handler to wait rather than sending SIGTERM.
		if ctrlErr == nil && wasCtrlAltDel {
			exitCode = system.GracefulShutdown(system.ShutdownConfig{
				Pid:               pid,
				IsChild:           true,
				PreSignalHook:     func() bool { return false },
				GracefulTimeout:   2 * time.Second,
				KillTimeout:       100 * time.Millisecond,
				ExpectedStartTime: c.vm.ProcessStartTime,
			})
		}

		// Fallback: signal-based shutdown (SIGTERM → wait → SIGKILL)
		if exitCode == nil {
			exitCode = system.GracefulShutdown(system.ShutdownConfig{
				Pid:               pid,
				IsChild:           true,
				GracefulTimeout:   2 * time.Second,
				KillTimeout:       1 * time.Millisecond,
				ExpectedStartTime: c.vm.ProcessStartTime,
			})
		}
	} else {
		if force {
			system.KillProcess(pid)
		}
		// Force path: SIGKILL already sent (or we skip SIGTERM for non-force
		// if no API socket). Either way, use aggressive timeouts — the process
		// should die within milliseconds.
		exitCode = system.GracefulShutdown(system.ShutdownConfig{
			Pid:               pid,
			IsChild:           true,
			GracefulTimeout:   100 * time.Millisecond,
			KillTimeout:       100 * time.Millisecond,
			ExpectedStartTime: c.vm.ProcessStartTime,
		})
	}

	// Capture exit code if not already captured (non-blocking)
	if exitCode == nil {
		exitCode = system.CaptureExitCode(pid)
	}

	// Persist exit code to database
	if exitCode != nil {
		if err := c.repo.UpdateExitCode(ctx, c.vm.ID, *exitCode); err != nil {
			return err
		}
	}

	// Update status to STOPPED
	if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusStopped); err != nil {
		return err
	}
	return nil
}

// --- Pause ---
//
// Pause is idempotent — no-op if already paused.
func (c *Controller) Pause(ctx context.Context) error {
	name := c.vm.Name

	// No-op — already paused
	if c.vm.Status == model.VMStatusPaused {
		return nil
	}

	// Cannot pause from these states
	if c.vm.Status == model.VMStatusStarting {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is still starting — cannot pause (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusStopped {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is stopped — cannot pause (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusStopping {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is shutting down — cannot pause (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusError || c.vm.Status == model.VMStatusCrashed {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is in %s state — cannot pause (current state: %s)", name, c.vm.Status, c.vm.Status))
	}

	// Valid transition — must be RUNNING
	if c.vm.APISocketPath == "" {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' has no API socket enabled", name))
	}

	// APISocketPath is already a full path from DB
	apiSocket := c.vm.APISocketPath
	client := firecracker.NewClient(apiSocket)
	defer client.Close()

	if err := client.PauseVM(ctx); err != nil {
		return err
	}

	// Only update DB, not in-memory status.
	if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusPaused); err != nil {
		return err
	}
	return nil
}

// --- Resume ---
func (c *Controller) Resume(ctx context.Context) error {
	name := c.vm.Name

	// No-op — already in or moving toward target state (RUNNING)
	if c.vm.Status == model.VMStatusRunning || c.vm.Status == model.VMStatusStarting {
		return nil
	}

	// Error/crashed state
	if c.vm.Status == model.VMStatusError || c.vm.Status == model.VMStatusCrashed {
		return errs.New(
			errs.CodeVMStateInvalid,
			fmt.Sprintf(
				"VM '%s' is in %s state — remove and recreate (current state: %s)",
				name,
				c.vm.Status,
				c.vm.Status,
			),
		)
	}

	// Wrong direction — stopped
	if c.vm.Status == model.VMStatusStopped {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is stopped — use start() instead (current state: %s)", name, c.vm.Status))
	}

	// Wrong direction — shutting down
	if c.vm.Status == model.VMStatusStopping {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is shutting down — use start() after it stops (current state: %s)", name, c.vm.Status))
	}

	// Valid transition — must be PAUSED
	if c.vm.APISocketPath == "" {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' has no API socket enabled", name))
	}

	// APISocketPath is already a full path from DB
	apiSocket := c.vm.APISocketPath
	client := firecracker.NewClient(apiSocket)
	defer client.Close()

	if err := client.ResumeVM(ctx); err != nil {
		return err
	}

	// Only update DB, not in-memory status.
	if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusRunning); err != nil {
		return err
	}
	return nil
}

// --- Start ---
func (c *Controller) Start(ctx context.Context) error {
	name := c.vm.Name

	// Log nested virt status for observability
	if c.vm.NestedVirt {
		slog.Info("VM has nested virtualization enabled", "name", name)
	}

	// No-op — already in or moving toward target state (RUNNING),
	// or will be stopped soon (retry start after)
	if c.vm.Status == model.VMStatusRunning || c.vm.Status == model.VMStatusStarting ||
		c.vm.Status == model.VMStatusStopping {
		return nil
	}

	// Error/crashed state
	if c.vm.Status == model.VMStatusError || c.vm.Status == model.VMStatusCrashed {
		return errs.New(
			errs.CodeVMStateInvalid,
			fmt.Sprintf(
				"VM '%s' is in %s state — remove and recreate (current state: %s)",
				name,
				c.vm.Status,
				c.vm.Status,
			),
		)
	}

	// Wrong direction — paused
	if c.vm.Status == model.VMStatusPaused {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is paused — use resume() instead (current state: %s)", name, c.vm.Status))
	}

	// Valid transition — must be STOPPED
	if c.vm.APISocketPath == "" {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' has no API socket enabled", name))
	}

	// APISocketPath is already a full path from DB
	apiSocket := c.vm.APISocketPath
	client := firecracker.NewClient(apiSocket)
	defer client.Close()

	if _, err := client.StartInstance(ctx); err != nil {
		return err
	}

	// Only update DB, not in-memory status.
	if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusRunning); err != nil {
		return err
	}
	return nil
}

// --- Reboot ---
func (c *Controller) Reboot(ctx context.Context, force bool) error {
	if err := c.Stop(ctx, force); err != nil {
		return err
	}
	return c.Start(ctx)
}

// --- SnapshotCreate ---
// Creates a Firecracker snapshot. If pauseOnly is true and the VM was running,
// it stays paused after the snapshot. If pauseOnly is false, the VM is resumed.
func (c *Controller) SnapshotCreate(ctx context.Context, cfg model.SnapshotCreateConfig) (err error) {
	name := c.vm.Name

	// Validate state — snapshot requires RUNNING or PAUSED
	if c.vm.Status == model.VMStatusStarting {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is still starting — cannot snapshot (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusStopped {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is stopped — cannot snapshot (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusStopping {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is shutting down — cannot snapshot (current state: %s)", name, c.vm.Status))
	}
	if c.vm.Status == model.VMStatusError || c.vm.Status == model.VMStatusCrashed {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("VM '%s' is in %s state — cannot snapshot (current state: %s)", name, c.vm.Status, c.vm.Status))
	}

	if c.vm.APISocketPath == "" {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("Socket not found for VM '%s'. Must be running with --enable-api-socket", name))
	}

	apiSocket := c.vm.APISocketPath
	client := firecracker.NewClient(apiSocket)
	wasRunning := c.vm.Status == model.VMStatusRunning

	defer func() {
		// Resume if we paused it and caller didn't want pause-only.
		if wasRunning && !cfg.PauseOnly {
			if resumeErr := client.ResumeVM(ctx); resumeErr != nil {
				if _, ok := errs.AsType[*errs.DomainError](resumeErr); ok {
					slog.Warn("Failed to resume VM after snapshot — leaving in paused state", "name", name)
				} else {
					err = resumeErr
				}
			} else if updateErr := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusRunning); updateErr != nil {
				if _, ok := errs.AsType[*errs.DomainError](updateErr); ok {
					slog.Warn("Failed to resume VM after snapshot — leaving in paused state", "name", name)
				} else {
					err = updateErr
				}
			} else {
				c.vm.Status = model.VMStatusRunning
			}
		}
		client.Close()
	}()

	// Pause before snapshotting (Firecracker requires VM to be paused)
	if wasRunning {
		if err := client.PauseVM(ctx); err != nil {
			return err
		}
		if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusPaused); err != nil {
			return err
		}
		c.vm.Status = model.VMStatusPaused
	}

	// PATCH drive to phantom path so the vmstate captures a snapshot-local
	// path instead of the source VM's path.
	if cfg.PhantomRootfsPath != "" {
		if patchErr := client.UpdateDrivePath(ctx, "rootfs", cfg.PhantomRootfsPath); patchErr != nil {
			return errs.WrapMsg(errs.CodeSnapshotCreateFailed,
				fmt.Sprintf("failed to patch drive to phantom path: %v", patchErr), patchErr)
		}
	}

	// Defer: restore original drive path after snapshot (runs before resume defer).
	if cfg.PhantomRootfsPath != "" {
		defer func() {
			if restoreErr := client.UpdateDrivePath(ctx, "rootfs", cfg.RootfsPath); restoreErr != nil {
				slog.Warn("Failed to restore drive path after snapshot",
					"vm", name, "error", restoreErr)
			}
		}()
	}

	_, snapshotErr := client.CreateSnapshot(ctx, cfg.MemFile, cfg.StateFile)
	if snapshotErr != nil {
		return snapshotErr
	}
	return nil
}

// --- SnapshotRestore ---
// Loads a snapshot into Firecracker (must already be spawned in snapshot mode
// with a ready API socket). Handles: LoadSnapshot with network override,
// vsock reconfiguration, and status update.
func (c *Controller) SnapshotRestore(ctx context.Context, cfg model.SnapshotRestoreConfig) error {
	if c.vm.APISocketPath == "" {
		return errs.New(errs.CodeVMStateInvalid,
			fmt.Sprintf("Socket not found for VM '%s'. Firecracker must be spawned before loading snapshot", c.vm.Name))
	}

	client := firecracker.NewClient(c.vm.APISocketPath)
	defer client.Close()

	// Map the VM's TAP device to eth0 via network_overrides so Firecracker
	// doesn't look for the original TAP name from the snapshot vmstate.
	if _, err := client.LoadSnapshot(
		ctx,
		cfg.MemFile,
		cfg.StateFile,
		cfg.Resume,
		cfg.NetworkOverrides,
		cfg.VsockUDSPath,
	); err != nil {
		return err
	}

	// NOTE: No PutVsock call needed here. vsock_override in LoadSnapshot
	// already sets the correct host-side UDS path. The guest CID is preserved
	// from the vmstate and matches what the guest agent uses. PutVsock is
	// pre-boot only — calling it on a resumed VM would fail or reset the
	// vsock device, breaking the guest agent's existing connection.

	// Update status
	if cfg.Resume {
		if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusRunning); err != nil {
			return err
		}
		c.vm.Status = model.VMStatusRunning
	} else {
		if err := c.repo.UpdateStatus(ctx, c.vm.ID, model.VMStatusPaused); err != nil {
			return err
		}
		c.vm.Status = model.VMStatusPaused
	}
	return nil
}

// --- AttachVolume ---
// AttachVolume hotplugs a drive into the running Firecracker process and
// persists the config so it survives reboot.
func (c *Controller) AttachVolume(ctx context.Context, driveConfig model.DriveConfig) error {
	apiSocket := c.vm.APISocketPath
	configPath := c.vm.ConfigPath

	// Hotplug into the running Firecracker process
	client := firecracker.NewClient(apiSocket)
	err := client.PutDrive(ctx, driveConfig)
	client.Close()
	if err != nil {
		return err
	}

	// Persist to the Firecracker config JSON so it survives reboot
	configMgr := NewFirecrackerConfigManager(configPath)
	if err := configMgr.AddDrive(driveConfig); err != nil {
		return err
	}

	slog.Info("Attached volume", "drive_id", driveConfig.DriveID, "vm", c.vm.Name)
	return nil
}

// --- DetachVolume ---
// DetachVolume hot-unplugs a drive from the running Firecracker process and
// removes it from the persisted config.
func (c *Controller) DetachVolume(ctx context.Context, driveID string) error {
	apiSocket := c.vm.APISocketPath
	configPath := c.vm.ConfigPath

	// Call Firecracker API to delete the drive
	client := firecracker.NewClient(apiSocket)
	err := client.DeleteDrive(ctx, driveID)
	client.Close()
	if err != nil {
		return err
	}

	// Update Firecracker config JSON on disk
	configMgr := NewFirecrackerConfigManager(configPath)
	if _, err := configMgr.RemoveDrive(driveID); err != nil {
		return err
	}

	slog.Info("Detached volume", "drive_id", driveID, "vm", c.vm.Name)
	return nil
}
